jaxlib/mosaic/gpu/custom_call.cc

```diff
@@ -89,8 +92,82 @@ namespace {
 using MosaicInitFunc = void(void****);
 using MosaicHostFunc = void(void**);
 
+absl::StatusOr<std::pair<std::string, std::string>> GetSmAndPtxIsaVersion() {
+  // Assumes driver has been initialized and a context exists. XLA already has
+  // some utilities to query this, but we try to stay runtime-agnostic, so we
+  // build our own here.
+  CUdevice device;
+  if (cuCtxGetDevice(&device) != CUDA_SUCCESS) {
+    return absl::InternalError("Failed to get device for current context");
+  }
+  int major = 0;
+  if (cuDeviceGetAttribute(&major, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR,
+                           device) != CUDA_SUCCESS) {
+    return absl::InternalError("Failed to get major compute capability");
+  }
+  int minor = 0;
+  if (cuDeviceGetAttribute(&minor, CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MINOR,
+                           device) != CUDA_SUCCESS) {
+    return absl::InternalError("Failed to get minor compute capability");
+  }
+  // "base" compute capability as reported by the driver.
+  // For example for a Hopper H200 GPU this would return sm_90, and never
+  // sm_90a.
+  std::string sm_base = "sm_" + std::to_string(major) + std::to_string(minor);
+
+  const std::string triple = "nvptx64-nvidia-cuda";
+  std::string error;
+  const llvm::Target* target =
+      llvm::TargetRegistry::lookupTarget(triple, error);
+  if (target == nullptr) {
+    return absl::InternalError(absl::StrFormat(
+        "Failed to lookup LLVM target based on triple %s: %s", triple, error));
+  }
+
+  // Check if there's a variant of the current SM that ends in "a"
+  // (has architecture-specific capabilities)
+  const char* sm_arch_specific = nullptr;
+  {
+    // generic subtarget
+    const llvm::MCSubtargetInfo* subtarget_info =
+        target->createMCSubtargetInfo(triple, "", "");
+    if (subtarget_info == nullptr) {
+      return absl::InternalError(absl::StrFormat(
+          "Failed to get generic LLVM subtarget info for triple %s", triple));
+    }
+    for (const llvm::SubtargetSubTypeKV& subtype :
+         subtarget_info->getAllProcessorDescriptions()) {
+      if (absl::StartsWith(subtype.Key, sm_base) &&
+          absl::EndsWith(subtype.Key, "a")) {
+        sm_arch_specific = subtype.Key;
+        break;
+      }
+    }
+  }
+
+  const std::string sm = sm_arch_specific ? sm_arch_specific : sm_base;
+
+  const llvm::MCSubtargetInfo* subtarget_info =
+      target->createMCSubtargetInfo(triple, sm, "");
+  if (subtarget_info == nullptr) {
+    return absl::InternalError(
+        absl::StrFormat("Failed to get LLVM subtarget info for sm %s", sm));
+  }
+
+  for (const llvm::SubtargetFeatureKV& feature :
+       subtarget_info->getEnabledProcessorFeatures()) {
+    if (absl::StartsWith(feature.Key, "ptx")) {
+      std::string ptx_isa = feature.Key;
+      return std::pair<std::string, std::string>(sm, ptx_isa);
```

Comment:
you should be able to `std::make_pair` to avoid the extra annotations I think

Reply:
Thank you, fixed in the latest push.